include/git2/branch.h

```diff
@@ -169,116 +166,127 @@ GIT_EXTERN(int) git_branch_lookup(
+ *
+ * @return 0 on success; GIT_ENOTFOUND if there's no branch named `branch_name`
+ *         or an error code
  */
-GIT_EXTERN(int) git_branch_set_upstream(git_reference *branch, const char *upstream_name);
+GIT_EXTERN(int) git_branch_set_upstream(
+	git_reference *branch,
+	const char *branch_name);
 
 /**
- * Return the name of the reference supporting the remote tracking branch,
- * given the name of a local branch reference.
- *
- * @param out Pointer to the user-allocated git_buf which will be
- * filled with the name of the reference.
+ * Get the upstream name of a branch
  *
- * @param repo the repository where the branches live
+ * Given a local branch, this will return its remote-tracking branch information,
+ * as a full reference name, ie. "feature/nice" would become
+ * "refs/remote/origin/feature/nice", depending on that branch's configuration.
  *
+ * @param out The buffer into which the name will be written.
+ * @param repo the repository where the branches live.
  * @param refname reference name of the local branch.
  *
- * @return 0, GIT_ENOTFOUND when no remote tracking reference exists,
- *     otherwise an error code.
+ * @return 0 on success, GIT_ENOTFOUND when no remote tracking reference exists,
+ *         or an error code.
  */
 GIT_EXTERN(int) git_branch_upstream_name(
 	git_buf *out,
 	git_repository *repo,
 	const char *refname);
 
 /**
- * Determine if the current local branch is pointed at by HEAD.
+ * Determine if HEAD points to the given branch
  *
- * @param branch Current underlying reference of the branch.
+ * @param branch A reference to a local branch.
  *
- * @return 1 if HEAD points at the branch, 0 if it isn't,
- * error code otherwise.
+ * @return 1 if HEAD points at the branch, 0 if it isn't, or a negative value
+ * as an error code.
  */
 GIT_EXTERN(int) git_branch_is_head(
 	const git_reference *branch);
 
 /**
- * Determine if the current branch is checked out in any linked
- * repository.
+ * Determine if any HEAD points to the current branch
+ *
+ * This will iterate over all known linked repositories (usually in the form of
+ * worktrees) and report whether their HEADs is pointing at the current branch.
  *
- * @param branch Reference to the branch.
+ * @param branch A reference to a local branch.
  *
- * @return 1 if branch is checked out, 0 if it isn't,
- * error code otherwise.
+ * @return 1 if branch is checked out, 0 if it isn't, an error code otherwise.
  */
 GIT_EXTERN(int) git_branch_is_checked_out(
 	const git_reference *branch);
 
 /**
- * Return the name of remote that the remote tracking branch belongs to.
+ * Extract the remote name of a branch
  *
- * @param out Pointer to the user-allocated git_buf which will be filled with the name of the remote.
+ * This will returns the name of the remote a remote-tracking branch belongs to.
+ * IOW, it extracts "test" from "refs/remotes/test/master", checks all remotes
+ * for a compatible refspec, and returns the first match.
```

Comment:
This one is the one I was most surprised by — I didn't expect a round-trip through the list of remotes to happen here…

I'm tempted to provide a lower-level version under that name and rename this one to be less "accessor"-y, so that you can get the name quickly-ish. Opinions ?

Reply:
At first this does sound weird to me, too. But you could have a remote that isn't named anything like what the refspecs will do. E.g. the remote "foobar" could definitely fetch into "barfoo" if the refspec is configured like that.

So we definitely need this roundtrip for correctness, but I wonder about extracting "test" first. Shouldn't we use the reference as-is and directly check it against refspecs, returning the first remote with a matching one?

Reply:
And by the way, you document the function to return `GIT_EAMBIGUOUS` in there are several matching refspecs, which sounds reasonable to me. But here in the description you say that we're returning the first matching remote, which directly contradicts that.

Reply:
I'm not sure about the roundtrips. IIRC reading the code left me with the feeling that a branch has a "belongs to" relation to its remote (hence the `"." means "the current repository's branches"` comment), and that some of those roundtrips aren't actually correct. For example, you can load a branch using its underlying reference name, but a subsequent operation with it will fail because this one will notice an ambiguity, or the configuration is somehow "broken" — the "branch.*" configuration key has been renamed, or its refspecs changed; I've been guilty of doing both, and I know GitX trusts the refdb for its remote name by reading `git for-each-ref` instead of doing this, hence all the suspicion 😉.